### PR TITLE
*: refactor API util functions

### LINKED ISSUE #4728: Merge `server/api/util` and `pkg/apiutil`
## Enhancement Task
We can merge them because they all belong to utility functions of API.

### LINKED ISSUE #4741: api_test: Some tests do useless check options when doing json request
## Enhancement Task
In test code of package `server/api `,  we use a lot of function `PostJSON`. `PostJSON` will check whether response code is StatusOK. If not, it returns error. At the same time, `PostJSON` has `checkOpt` params which will do some checks.

Firstly, `PostJSON` returns error and don't process check option when response code is not StatusOK. This is not in line with people's usage habits. There are codes where this is used in PD, so in fact these codes are useless.
Such as https://github.com/tikv/pd/blob/1a353ae2058d75b322afe1ef7783e4b441adaddc/server/api/admin_test.go#L146-L147

Secondly, `checkOpt` is used to test. So it is not appropriate to put it in `util.go/api_util.go`. We should put it only in the test file.

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: Close #4741 Close #4728

### What is changed and how it works?
remove server/api/util.go
refactor pkg/apiutil
add util for API in testutil

```commit-message
Merge server/api/util and pkg/apiutil and refactor apiutil
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/testutil/api_check.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/pingcap/check"
+	"github.com/tikv/pd/pkg/apiutil"
+)
+
+type APICheckerUtil struct {
+	c *check.C
+}
```
Comment (rleungx) [inline pkg/testutil/api_check.go:29]: Maybe use check directly in the function?

File: pkg/apiutil/apiutil.go [hunk 2]
```diff
@@ -241,12 +183,26 @@ func PatchJSON(client *http.Client, url string, data []byte) (*http.Response, er
...
 }
 
-// PostJSONIgnoreResp
+// PostJSONIgnoreResp is used to do post requeset with JSON body and ignore response.
 func PostJSONIgnoreResp(client *http.Client, url string, data []byte) error {
 	resp, err := PostJSON(client, url, data)
 	return checkResponse(resp, err)
 }
 
+// DoDelete
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:192]: need more comments

File: pkg/apiutil/apiutil.go [hunk 3]
```diff
@@ -241,12 +183,26 @@ func PatchJSON(client *http.Client, url string, data []byte) (*http.Response, er
 	return client.Do(req)
 }
 
-// PostJSONIgnoreResp
+// PostJSONIgnoreResp is used to do post requeset with JSON body and ignore response.
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:186]:
```suggestion
// PostJSONIgnoreResp is used to do post request with JSON body and ignore response.
```

File: pkg/apiutil/apiutil.go [hunk 4]
```diff
@@ -232,7 +174,7 @@ func GetJSON(client *http.Client, url string, data []byte) (*http.Response, erro
 	return client.Do(req)
 }
 
-// PatchJSON
+// PatchJSON is used to do patch requeset
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:177]:
```suggestion
// PatchJSON is used to do patch request
```

File: pkg/apiutil/parseutil.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 TiKV Project Authors.
```
Comment (rleungx) [inline pkg/apiutil/parseutil.go:1]: Can we just put these functions into apiutil.go?

File: pkg/apiutil/apiutil.go (hunk 2 above)
Comment (CabinfeverB) [reply pkg/apiutil/apiutil.go]: added

File: pkg/testutil/api_check.go (hunk 1 above)
Comment (CabinfeverB) [reply pkg/testutil/api_check.go]: refactor these functions

File: pkg/apiutil/apiutil.go [hunk 6]
```diff
@@ -183,13 +188,13 @@ func PatchJSON(client *http.Client, url string, data []byte) (*http.Response, er
...
 
-// PostJSONIgnoreResp is used to do post requeset with JSON body and ignore response.
+// PostJSONIgnoreResp is used to do post request with JSON body and ignore response.
 func PostJSONIgnoreResp(client *http.Client, url string, data []byte) error {
 	resp, err := PostJSON(client, url, data)
 	return checkResponse(resp, err)
 }
 
-// DoDelete
+// DoDelete is used to send delete requeset and return http response code.
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:197]: request

File: server/api/operator_test.go [hunk 7]
```diff
@@ -267,7 +265,7 @@ func (s *testTransferRegionOperatorSuite) TestTransferRegionWithPlacementRule(c
 			expectSteps:         "",
 		},
 		{
-			name:                "customized placement rule with invalid peer role",
+			name:                "tustomized placement rule with invalid peer role",
```
Comment (rleungx) [inline server/api/operator_test.go:268]: change them back

File: server/api/scheduler.go [hunk 8]
```diff
@@ -325,5 +325,5 @@ func (h *schedulerConfigHandler) GetSchedulerConfig(w http.ResponseWriter, r *ht
 		sh.ServeHTTP(w, r)
 		return
 	}
-	h.rd.JSON(w, http.StatusNotAcceptable, errNoImplement.Error())
```
Comment (CabinfeverB) [inline server/api/scheduler.go:328]: The condition to process this is that RaftCluster is nil, so it was wrong before.

File: pkg/apiutil/apiutil.go (hunk 6 above)
Comment (CabinfeverB) [reply pkg/apiutil/apiutil.go]: fix

File: server/api/operator_test.go (hunk 7 above)
Comment (CabinfeverB) [reply server/api/operator_test.go]: fix

File: pkg/testutil/api_check.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+func StatusOK(c *check.C) func([]byte, int) {
+	return Status(c, http.StatusOK)
+}
+
+// StatusNotOK is used to check whether http response code is not equal http.StatusOK
+func StatusNotOK(c *check.C) func([]byte, int) {
+	return func(_ []byte, i int) {
+		c.Assert(i == http.StatusOK, check.IsFalse)
+	}
+}
```
Comment (rleungx) [inline pkg/testutil/api_check.go:44]: Is it duplicated with `Status`?

File: server/api/admin_test.go [hunk 10]
```diff
@@ -91,10 +91,10 @@ func (s *testAdminSuite) TestDropRegion(c *C) {
 
 func (s *testAdminSuite) TestPersistFile(c *C) {
 	data := []byte("#!/bin/sh\nrm -rf /")
-	err := postJSON(testDialClient, s.urlPrefix+"/admin/persist-file/fun.sh", data)
-	c.Assert(err, NotNil)
+	err := tu.CheckPostJSON(testDialClient, s.urlPrefix+"/admin/persist-file/fun.sh", data, tu.StatusNotOK(c))
+	c.Assert(err, IsNil)
```
Comment (rleungx) [inline server/api/admin_test.go:95]: Why change the assert?

File: pkg/errs/errno.go [hunk 11]
```diff
@@ -87,6 +87,7 @@ var (
 	ErrCacheOverflow                    = errors.Normalize("cache overflow", errors.RFCCodeText("PD:scheduler:ErrCacheOverflow"))
 	ErrInternalGrowth                   = errors.Normalize("unknown interval growth type error", errors.RFCCodeText("PD:scheduler:ErrInternalGrowth"))
 	ErrSchedulerCreateFuncNotRegistered = errors.Normalize("create func of %v is not registered", errors.RFCCodeText("PD:scheduler:ErrSchedulerCreateFuncNotRegistered"))
+	ErrSchedulerConfigNoImplement       = errors.Normalize("schedulers config no implement", errors.RFCCodeText("PD:scheduler:ErrSchedulerConfigNoImplement"))
```
Comment (rleungx) [inline pkg/errs/errno.go:90]: Where do we use it?

File: server/api/admin_test.go (hunk 10 above)
Comment (CabinfeverB) [reply server/api/admin_test.go:95]: > Why change the assert?

`postJSON` will return err if `StatusCode` is not equal 200. But now this check does not exist in `CheckPostJSON `.  If need, you can add check action `StatusNotOK`.

File: pkg/testutil/api_check.go (hunk 9 above)
Comment (CabinfeverB) [reply pkg/testutil/api_check.go:44]: > Is it duplicated with `Status`?

Easy to use, so the user doesn't have to add code `200` every time.

File: pkg/errs/errno.go (hunk 11 above)
Comment (CabinfeverB) [reply pkg/errs/errno.go:90]: Can remove it. In the previous impl, `schedulerConfigHandler.GetSchedulerConfig` has an `errNoImplement` which is defined in `/server/api/util.go`. But `GetSchedulerConfigHandler` returns nil only if `RaftCluster` is nil, so this error output is incorrect. So I don't use ErrSchedulerConfigNoImplement.

**Review by rleungx — APPROVED** [review]
The rest LGTM.

**Review by nolouch — APPROVED** [review]
lgtm

**Comment by nolouch** [comment]
@CabinfeverB  CI failed

**Comment by CabinfeverB** [comment]
> @CabinfeverB CI failed

There was some conflict. It's been resolved now.

**Pull request merged**